### PR TITLE
[IMP] auth_ldap - Allow several '%s' occurences in the login filter.

### PR DESCRIPTION
**Pull request by azmeuk**
### Description of the issue/feature this PR addresses:

This patch allows to match login on several LDAP attributes. This is very useful if you want to login either against an email or an uid.
Then you can use a filter such as:

```
   (|(mail=%s)(uid=%s))
```

### Current behavior before PR:

A filter like `(|(mail=%s)(uid=%s))` raises a `TypeError`

### Desired behavior after PR is merged:

A filter like `(|(mail=%s)(uid=%s))` is usable.

Closes #25502
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by azmeuk** [comment]
@beledouxdenis @tivisse @d-fence @hbrunn It would hugely appreciate if you could find a few minutes to review this patch. This should be pretty straightforward :)

File: addons/auth_ldap/models/res_company_ldap.py [hunk 1]
```diff
@@ -79,19 +79,18 @@ def _connect(self, conf):
...
-
-            # Get rid of (None, attrs) for searchResultReference replies
-            results = [i for i in results if i[0]]
-            if len(results) == 1:
-                entry = results[0]
-                dn = results[0][0]
+        dn, entry = False, False
+        filter = filter_format(
+            conf['ldap_filter'],
+            (login,) * conf['ldap_filter'].count('%s')
```
Comment (Yajo) [inline addons/auth_ldap/models/res_company_ldap.py:85]: You can kill this with just setting `%%s` in your filter, although I'm not sure why would you want to do that.

Instead, it'd be better to allow formatting using a dict, so you can use as many occurrences of `%(login)s` as you need, or you could even add things such as `%(email)s` also.

A backwards-compatibility layer could be added too. I'm not sure if this should go for master though.

Comment (azmeuk) [reply addons/auth_ldap/models/res_company_ldap.py:85]: Thank you for your review.

I agree that a `%(login)s` based solution would be the best. Here I wanted to keep the patch as small as possible, but I am ok to provide another more complete patch. If not in master, where should this be implemented in your opinion?

As for this patch, what would you need me to do to approve it?

Comment (azmeuk) [reply addons/auth_ldap/models/res_company_ldap.py:85]: @Yajo Would you consider approving this patch?

Comment (Yajo) [reply addons/auth_ldap/models/res_company_ldap.py:85]: It's just a comment, my approval here is pointless, you have to convince Odoo people 😅

Comment (azmeuk) [reply addons/auth_ldap/models/res_company_ldap.py:85]: I thought two approvals from anyone were required, but maybe this just applies for OCA.

Comment (pedrobaeza) [reply addons/auth_ldap/models/res_company_ldap.py:85]: Yes, that is OCA, not here

**Comment by xmo-odoo** [comment]
I don't really know much about ldap, but shouldn't the warning about having at least one format placeholder be left in? I don't know if there's a use-case for allowing a completely hard-coded filter.

**Comment by azmeuk** [comment]
@xmo-odoo You are right, this is technically valid but does not really make sense, except maybe for debugging purpose. I added a warning.

**Review by xmo-odoo — COMMENTED** [review]
Hey @azmeuk sorry for the silence, after discussing it internally we're OK with adding this feature. @Yajo's suggestion was a concern but:

* python-ldap doesn't support mappings in `filter_format` so we'd have to recode that (though it's not super hard)
* we'd need a user-case for having multiple values (what would they be? the callers have essentially nothing else since this is the auth process)
* either way we'd have to handle the dict-based and seq-based versions just for backwards compatibility

However @odony would like the `ldap_filter` field to have some explanations as currently it's a bit... terse, for people who don't already know ldap well. Do you feel like doing that or should we take it up internally?

**Comment by azmeuk** [comment]
> Hey @azmeuk sorry for the silence, after discussing it internally we're OK with adding this feature.
Great news!

> However @odony would like the `ldap_filter` field to have some explanations as currently it's a bit... terse, for people who don't already know ldap well. Do you feel like doing that or should we take it up internally?

I am sorry I won't have much time to spend on this in the coming weeks.

File: addons/auth_ldap/models/res_company_ldap.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 from ldap.filter import filter_format
 
 from odoo import _, api, fields, models, tools
-from odoo.exceptions import AccessDenied
+from odoo.exceptions import AccessDenied, ValidationError
```
Comment (mart-e) [inline addons/auth_ldap/models/res_company_ldap.py:9]: unused import

Comment (xmo-odoo) [reply addons/auth_ldap/models/res_company_ldap.py]: Fixed :+1: (I'd added a "hard" constraint on the presence of placeholders but after reflexion that seemed a bit too extreme, but I forgot to remove the import).

**Comment by azmeuk** [comment]
@xmo-odoo there has been a few weeks since activity. Is there something more that I can help with to merge this PR?

**Review by odony — APPROVED** [review]
Tested, works for me. And we improve the tooltip and cleanup/simplify the code at the same time, great! :+1: 

@azmeuk and @xmo-odoo thank you for the nice work!

@robodoo r+

**Comment by KangOl** [comment]
@robodoo rebase-ff

**Pull request closed without merging**